File: svc-event/src/test/java/com/beancounter/event/TestMsftFlow.java
package com.beancounter.event;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import com.beancounter.auth.common.TokenService;
import com.beancounter.client.services.PortfolioServiceClient;
import com.beancounter.common.contracts.PortfoliosResponse;
import com.beancounter.common.contracts.PositionResponse;
import com.beancounter.common.event.CorporateEvent;
import com.beancounter.common.input.TrustedEventInput;
import com.beancounter.common.input.TrustedTrnEvent;
import com.beancounter.common.input.TrustedTrnQuery;
import com.beancounter.common.model.Portfolio;
import com.beancounter.common.utils.DateUtils;
import com.beancounter.event.integration.PositionGateway;
import com.beancounter.event.service.EventService;
import com.beancounter.event.service.PositionService;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Collection;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.core.io.ClassPathResource;
import org.springframework.test.context.ActiveProfiles;

@SpringBootTest(classes = EventBoot.class)
@ActiveProfiles("test") // Ignore Kafka
public class TestMsftFlow {
  @Autowired
  private EventService eventService;

  @Autowired
  private PositionService positionService;

  @Test
  void is_DividendRequestIgnore() throws Exception {
    ObjectMapper om = new ObjectMapper();

    TrustedEventInput trustedEvent = om.readValue(
        new ClassPathResource("/msft-flow/1-event.json").getFile(),
        TrustedEventInput.class);
    PortfoliosResponse whereHeld = om.readValue(
        new ClassPathResource("/msft-flow/2-where-held.json").getFile(), PortfoliosResponse.class);
    PositionResponse positionResponse = om.readValue(
        new ClassPathResource("/msft-flow/3-position.json").getFile(), PositionResponse.class);

    PositionGateway positionGateway = Mockito.mock(PositionGateway.class);
    CorporateEvent dividend = trustedEvent.getData();
    DateUtils dateUtils = new DateUtils();

    when(
        positionGateway.get("demo",
            dividend.getAssetId(),
            dateUtils.getDateString(dividend.getRecordDate()))
    ).thenReturn(positionResponse);

    Portfolio portfolio = whereHeld.getData().iterator().next();

    when(
        positionGateway
            .query(
                "demo",
                TrustedTrnQuery.builder()
                    .portfolio(portfolio)
                    .tradeDate(dividend.getRecordDate())
                    .assetId(dividend.getAssetId())
                    .build())
    ).thenReturn(positionResponse);
    PortfolioServiceClient.PortfolioGw portfolioGw = mock(PortfolioServiceClient.PortfolioGw.class);
    when(portfolioGw.getWhereHeld(
        "demo",
        dividend.getAssetId(),
        dateUtils.getDateString(dividend.getRecordDate())))
        .thenReturn(whereHeld);

    TokenService tokenService = mock(TokenService.class);
    when(tokenService.getBearerToken()).thenReturn("demo");
    PortfolioServiceClient portfolioServiceClient =
        new PortfolioServiceClient(portfolioGw, tokenService);

    positionService.setPositionGateway(positionGateway);
    positionService.setTokenService(tokenService);
    positionService.setPortfolioClientService(portfolioServiceClient);

    Collection<TrustedTrnEvent> results = eventService.processMessage(trustedEvent);
    assertThat(results).isEmpty();

  }
}

File: svc-event/src/test/kotlin/com/beancounter/event/TestMsftFlow.kt
package com.beancounter.event

import com.beancounter.auth.AutoConfigureMockAuth
import com.beancounter.auth.MockAuthConfig
import com.beancounter.auth.TokenService
import com.beancounter.client.services.PortfolioServiceClient
import com.beancounter.common.contracts.PortfoliosResponse
import com.beancounter.common.contracts.PositionResponse
import com.beancounter.common.input.TrustedEventInput
import com.beancounter.common.input.TrustedTrnQuery
import com.beancounter.common.utils.BcJson.Companion.objectMapper
import com.beancounter.event.integration.PositionGateway
import com.beancounter.event.service.EventService
import com.beancounter.event.service.PositionService
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.Mockito.`when`
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.core.io.ClassPathResource
import org.springframework.security.oauth2.jwt.JwtDecoder
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.bean.override.mockito.MockitoBean

/**
 * Simple flow of various corporate events for Microsoft.
 */
@SpringBootTest
@ActiveProfiles("test")
@AutoConfigureMockAuth
class TestMsftFlow {
    @MockitoBean
    private lateinit var jwtDecoder: JwtDecoder

    @MockitoBean
    private lateinit var positionGateway: PositionGateway

    @Autowired
    private lateinit var eventService: EventService

    @Autowired
    private lateinit var positionService: PositionService

    @Autowired
    private lateinit var mockAuthConfig: MockAuthConfig

    @Test
    fun is_DividendRequestIgnoreDueToZeroHolding() {
        assertThat(mockAuthConfig).isNotNull
        assertThat(positionGateway).isNotNull

        val trustedEvent =
            objectMapper.readValue(
                ClassPathResource("/msft-flow/1-event.json").file,
                TrustedEventInput::class.java
            )
        val whereHeld =
            objectMapper.readValue(
                ClassPathResource("/msft-flow/2-where-held.json").file,
                PortfoliosResponse::class.java
            )
        val positionResponse =
            objectMapper.readValue(
                ClassPathResource("/msft-flow/3-position.json").file,
                PositionResponse::class.java
            )
        val mockPositionGateway = Mockito.mock(PositionGateway::class.java)
        val (_, _, _, assetId, recordDate) = trustedEvent.data

        `when`(
            mockPositionGateway["demo", assetId, recordDate.toString()]
        ).thenReturn(positionResponse)

        val portfolio = whereHeld.data.iterator().next()
        `when`(
            mockPositionGateway.query(
                "demo",
                TrustedTrnQuery(
                    portfolio,
                    recordDate,
                    assetId
                )
            )
        ).thenReturn(positionResponse)

        val mockPortfolioServiceClient = Mockito.mock(PortfolioServiceClient::class.java)
        `when`(mockPortfolioServiceClient.getWhereHeld(assetId, recordDate))
            .thenReturn(whereHeld)

        val tokenService = Mockito.mock(TokenService::class.java)
        `when`(tokenService.bearerToken)
            .thenReturn("demo")

        positionService.setTokenService(tokenService)
        positionService.setPortfolioClientService(mockPortfolioServiceClient)
        val results = eventService.process(trustedEvent)
        assertThat(results).isEmpty()
    }
}